pallets/gear/src/internal.rs

```diff
@@ -632,22 +674,50 @@ where
                     Self::remove_gas_reservation_with_task(dispatch.source(), reservation_id);
                 }
             }
-        }
 
-        if !dispatch.value().is_zero() {
-            // Reserving value from source for future transfer or unreserve.
-            CurrencyOf::<T>::reserve(&from, value)
-                .unwrap_or_else(|e| unreachable!("Unable to reserve requested value {:?}", e));
+            // `HoldBound` cost builder.
+            let hold_builder = HoldBound::<T>::by(CostsPerBlockOf::<T>::dispatch_stash());
+
+            // Calculating correct hold bound to lock gas.
+            let maximal_hold = hold_builder.clone().maximum_for_message(dispatch.id());
+            let bn_delay = delay.saturated_into::<BlockNumberFor<T>>();
+            let hold = hold_builder.duration(bn_delay).min(maximal_hold);
+
+            interval_finish = hold.expected();
+
+            // Locking funds for holding.
+            GasHandlerOf::<T>::lock(dispatch.id(), hold.lock())
+                .unwrap_or_else(|e| unreachable!("GasTree corrupted! {:?}", e));
+
+            if hold.expected_duration().is_zero() {
+                unreachable!("Hold duration cannot be zero");
+            }
+
+            if !dispatch.value().is_zero() {
+                // Reserving value from source for future transfer or unreserve.
+                CurrencyOf::<T>::reserve(&from, value)
+                    .unwrap_or_else(|e| unreachable!("Unable to reserve requested value {:?}", e));
+            }
```

Comment:
reserve of the value should be out of if-else branching

Reply:
done